File: Unit2-05/Vehicle.ts
/*
* Class that defines a vehicle.
*
* @author  Infinity de Guzman
* @version 1.0
* @since   2024-05-05
*/

export default class Vehicle {
    // variables
    private _colour: string
    private _licensePlate: string
    readonly numberOfDoors: number
    readonly maxSpeed: number
    private _speed: number
   
    /*
    * Constructor
    */
    // apparently, it's good practice to reclassify the variables in the constructor's parameters
    constructor(colour: string, licensePlate: string, numberOfDoors: number, maxSpeed: number) {
      this._colour = colour
      this._licensePlate = licensePlate
      this.numberOfDoors = numberOfDoors
      this.maxSpeed = maxSpeed
      this._speed = 0
    }
   
    /*
    * This method gets the colour
    */
    public get colour() {
      return this._colour
    }
   
    /*
    * This method gets the license plate
    */
    public get licensePlate() {
      return this._licensePlate
    }
   
    /*
    * This method gets the speed
    */
    public get speed() {
      return this._speed
    }
   
    /*
    * This method sets the colour
    */
    public set colour(value: string) {
      this._colour = value
    }
   
    /*
    * This method sets the license plate
    */
    public set licensePlate(value: string) {
      this._licensePlate = value
    }
   
    /*
    * This method shows the status of the vehicle
    */
    public status() {
      console.log(" -> Speed: " + this._speed)
      console.log(" -> Max Speed: " + this.maxSpeed)
      console.log(" -> Number Of Doors: " + this.numberOfDoors)
      console.log(" -> License Plate: " + this._licensePlate)
      console.log(" -> Colour: " + this._colour)
    }
   
    /*
    * This method changes the speed via the accelerating formula
    */
    public accelerate(accPower: number, accTime: number) {
      this._speed = (accPower * accTime) + this._speed
      // check if speed exceeds max speed
      if (this._speed > this.maxSpeed) {
        this._speed = this.maxSpeed
      }
    }
   
    /*
    * This method changes the speed via the braking formula
    */
    public brake(breakPower: number, breakTime: number) {
      this._speed -= breakPower * breakTime
      // check if speed is less than 0, because speed can't be negative!
      if (this._speed < 0) {
        this._speed = 0
      }
    }
}

File: Unit2-05/java/Vehicle.java
/**
 * Vehicle class.
 *
 * <p>This class represents a vehicle with various attributes.
 * {@code null} is returned when no attribute is set.
 *
 * @author Infinity de Guzman
 * @version 1.0
 * @since 2024-05-09
 */
public class Vehicle {
    /**
     * The colour of the vehicle.
     */
    private String colour;

    /**
     * The license plate of the vehicle.
     */
    private String licensePlate;

    /**
     * The number of doors on the vehicle.
     */
    private final int numberOfDoors;

    /**
     * The maximum speed of the vehicle.
     */
    private final int maxSpeed;

    /**
     * The current speed of the vehicle.
     */
    private int speed;

    /**
     * Constructor.
     *
     * @param colour        the colour of the vehicle
     * @param licensePlate  the license plate of the vehicle
     * @param numberOfDoors the number of doors on the vehicle
     * @param maxSpeed      the maximum speed of the vehicle
     */
    public Vehicle(
        String colour,
        String licensePlate,
        int numberOfDoors,
        int maxSpeed
    ) {
        this.colour = colour;
        this.licensePlate = licensePlate;
        this.numberOfDoors = numberOfDoors;
        this.maxSpeed = maxSpeed;
        this.speed = 0;
    }

    /**
     * Gets the colour of the vehicle.
     *
     * @return the colour of the vehicle
     */
    public String getColour() {
        return colour;
    }

    /**
     * Gets the license plate of the vehicle.
     *
     * @return the license plate of the vehicle
     */
    public String getLicensePlate() {
        return licensePlate;
    }

    /**
     * Gets the speed of the vehicle.
     *
     * @return the speed of the vehicle
     */
    public int getSpeed() {
        return speed;
    }

    /**
     * Sets the colour of the vehicle.
     *
     * @param colour the new colour to set
     */
    public void setColour(String colour) {
        this.colour = colour;
    }

    /**
     * Sets the license plate of the vehicle.
     *
     * @param licensePlate the new license plate to set
     */
    public void setLicensePlate(String licensePlate) {
        this.licensePlate = licensePlate;
    }

    /**
     * Prints the status of the vehicle.
     */
    public void status() {
        System.out.println(" -> Speed: " + speed);
        System.out.println(" -> Max Speed: " + maxSpeed);
        System.out.println(" -> Number Of Doors: " + numberOfDoors);
        System.out.println(" -> License Plate: " + licensePlate);
        System.out.println(" -> Colour: " + colour);
    }

    /**
     * Accelerates the vehicle.
     *
     * @param accPower the acceleration power
     * @param accTime  the acceleration time
     */
    public void accelerate(int accPower, int accTime) {
        final int acceleration = accPower * accTime;
        if (speed + acceleration <= maxSpeed) {
            speed += acceleration;
        } else {
            speed = maxSpeed;
        }
    }

    /**
     * Applies brakes to the vehicle.
     *
     * @param brakePower the braking power
     * @param brakeTime  the braking time
     */
    public void brake(int brakePower, int brakeTime) {
        final int deceleration = brakePower * brakeTime;
        if (speed - deceleration >= 0) {
            speed -= deceleration;
        } else {
            speed = 0;
        }
    }
}

